File: web/static/js/habit.js
/*    Copyright (C) 2021  The Project TONA Authors
*
*    This program is free software: you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation, either version 3 of the License, or
*    (at your option) any later version.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

(function(window){
    function Habit (){
        var self = {};
        
        self.API = "/api/habit";

        self.AddHabit = function (event){
            
            var modal = event.target.parentNode.parentNode;           
            var name = modal.querySelector("#habit-name");

            if (name.value != ""){
                var data = {name: name.value};
                fetch(self.API, {
                    method: 'post',
                    body: JSON.stringify(data),
                    headers: Tona.SetHeaders(),
                }).then(response => response.json()).then(function(data){
                    if (data['ok']){
                        var payload = data['payload'];
                        var menu = document.querySelector("#habit-menu-list");
                        var li = document.createElement("li");
                        li.innerHTML = '<a href="/habit/'+payload['id']+'">'+payload['name']+'</a>';
                        menu.append(li);
                        Tona.Modal.Close()
                        name.value = ""
                    }else{
                        console.warn(data)
                    }
                }).catch(function(error){
                    console.error(error);
                });
            }
        }; 

        self.AddCheckin = function(event, habit_id){
            var doc =  event.target.parentNode
            var id = "habit-checkin-name";
            var name = doc.querySelector("#"+id);
            var simplemde = Tona.QueueSimpleMDE[id]
            if ( Tona.IsSimpleMDE(simplemde) ){
                name.value = simplemde.value();
            }
            
            if (name.value != "") {

                var date = new Date()
                var data = {
                    name: name.value, 
                    checkin: date.getFullYear()+"-"+ (date.getMonth()+1) +"-"+ date.getDate(),
                    habit_id: habit_id
                }
                
                fetch( self.API+"/checkin", {
                    method: 'post',
                    body: JSON.stringify(data),
                    headers: Tona.SetHeaders(),
                }).then(response => response.json()).then(function(data){
                    if (data['ok']){
                        name.value = ""
                        if ( Tona.IsSimpleMDE(simplemde) ){
                            Tona.QueueSimpleMDE[id].value("");
                        }
                        Tona.Notification(event, "Your chekin is added", type="info", interval=1000)
                        self.LoadCheckin(event, habit_id);
                    }else{
                        Tona.Notification(event, data['message'])
                    }
                }).catch(function(error){
                    console.error(error);
                });
            }
        }

        self.LoadCheckin = function(event, habit_id, offset=1){
            var history = document.querySelector("#habit-checkin");
            if (offset == 1){
                history.innerHTML = "";
            }
            
            fetch(self.API +"/checkin?habit_id="+habit_id+"&offset="+offset+"&limit=10", {
                method: 'get',
                headers: Tona.SetHeaders(),
            }).then(response => response.json()).then(function(data){
                if (data['ok'] == true){
                    var payload = data['payload'];                    
                    
                    if (payload.length > 0){
                        for ( a = 0 ; a  < payload.length; a++ ){
                            var checkin = payload[a]
                            var msg = `<article class="media">
                                <div class="media-content">
                                <div class="content">
                                    <p>
                                    `+marked(checkin['name'])+`
                                    <small><a>`+checkin['checkin']+`</a></small>
                                    </p>
                                </div>
                                </div>
                            </article>`
                            history.innerHTML += msg
                        }
                        if (offset != 1){
                            history.querySelector("#checkin-show-more").remove()
                        }
                        history.innerHTML +=`<a onclick="Habit.LoadCheckin(`+null+`,`+habit_id+`,`+(offset+1)+`)" id="checkin-show-more">Show More</a>`
                    }else{
                        var sm = history.querySelector("#checkin-show-more");
                        if (sm != null){
                            sm.remove()
                            history.innerHTML +=`<a onclick="Habit.LoadCheckin(`+null+`,`+habit_id+`,`+1+`)" id="checkin-show-more">Show less</a>`
                        }
                    }   
                }
                
            }).catch(function(error){
                console.error(error);
            });

        };

        return self;
    }

    if (typeof(window.Habit) === 'undefined'){
        window.Habit = Habit();
    }

})(window);

File: web/habit/habit.py
# -*- coding: utf-8 -*-
#    Copyright (C) 2021  The Project TONA Authors
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
import datetime
from flask import render_template, request, jsonify, flash, Blueprint
from tona.web.main import app
from tona.utils import HTTPResponse, str2int, format_datetime
from tona.models.habit import Habit
from tona.models.habit_checkin import HabitCheckin


habit_bp = Blueprint('habit_bp', __name__,
                        template_folder='templates',
                        static_folder='static', static_url_path='assets',
                        url_prefix='/habit')

habit_api_bp = Blueprint('habit_api_bp', __name__, url_prefix='/api/habit')

@habit_bp.route("/")
@habit_bp.route("/<habit_id>")
def habit(habit_id=""):

    is_today = False
    habits_aux = []

    habits = []
    is_habit = False
    habit = None

    habits = Habit.select().where(Habit.active == True).order_by(Habit.name.asc())

    if habit_id == "today" or habit_id == "":
        is_today = True
        today = datetime.date.today()
        weekday = format_datetime(today, fmt_out='%A')        
        checked = Habit.select().join(HabitCheckin).where(
            Habit.frequency == 'daily',
            Habit.every.contains(weekday.lower()),
            HabitCheckin.checkin == today
            )
        has_checkin = [ r.id for r in checked ]
        unchecked = Habit.select().where(
           Habit.frequency == 'daily',
           Habit.id.not_in( has_checkin)
        )
        habits_aux = {
            'unchecked': unchecked,
            'checked': checked
        }
    elif habit_id == "archive":
        pass
    else:
        try:
            habit_id = str2int(habit_id)
            habit = Habit.exists(habit_id)
            is_habit = True
        except Exception as e:
            flash(str(e))
    rt = render_template(
        "habit.html",
        habits_aux=habits_aux,
        habits=habits,
        habit=habit,
        is_habit=is_habit,
        is_today=is_today)
    return rt


@habit_api_bp.route("", methods=['POST', 'GET'])
def api_habit():
    resp = HTTPResponse()
    try:
        if request.method == 'POST':
            data = request.json
            resp.payload = Habit.add(**data).to_dict()
        else:
            offset = int(request.args.get('offset', 1))
            limit = int(request.args.get('limit', 10))
            rows = Habit.select().order_by(Habit.name.desc()).paginate(offset, limit)
            data = []
            for row in rows:
                data.append(row.to_dict())
            resp.payload = data
        resp.ok = True
        resp.code = 200
    except Exception as e:
        app.logger.error(e)
        resp.message = str(e)
    return jsonify(resp.to_dict()), resp.code


@habit_api_bp.route("/checkin", methods=['POST', 'GET'])
def api_habit_checkin():
    resp = HTTPResponse()
    try:
        if request.method == 'POST':
            data = request.json
            resp.payload = HabitCheckin.add(**data).to_dict()
        else:
            offset = int(request.args.get('offset', 1))
            limit = int(request.args.get('limit', 10))
            habit_id = int(request.args.get('habit_id', 0))
            data = []
            rows = []
            if habit_id:
                rows = HabitCheckin.select().where(
                    HabitCheckin.habit_id == habit_id,
                ).order_by(HabitCheckin.checkin.desc()).paginate(offset, limit)
            else:
                rows = HabitCheckin.select().order_by(HabitCheckin.checkin.desc()).paginate(offset, limit)
            for row in rows:
                data.append(row.to_dict())
            resp.payload = data
        resp.ok = True
        resp.code = 200
    except Exception as e:
        app.logger.error(e)
        resp.message = str(e)
    return jsonify(resp.to_dict()), resp.code